lib/Coffer/Directory.hs

```diff
@@ -0,0 +1,94 @@
+module Coffer.Directory
+  ( Directory(..)
+  , emptyDir
+  , singleton
+  , allEntries
+  , mapMaybe
+  , mapEntry
+  , mapDir
+  , insertEntry
+  , trimEmptyDirs
+  , filterEntries
+  ) where
+
+import Control.Lens
+import qualified Data.HashMap.Strict as HashMap
+import Data.HashMap.Strict (HashMap)
+import qualified Data.Maybe as Maybe
+import Coffer.Path (PathSegment, entryPathParentDir, unPath)
+import Entry (Entry)
+import qualified Entry as E
+
+----------------------------------------------------------------------------
+-- Directory
+----------------------------------------------------------------------------
+
+data Directory = Directory
+  { _entries :: [Entry]
+  , _subdirs :: HashMap PathSegment Directory
+  }
+  deriving stock (Show, Eq)
```

Comment:
Why `Directory` doesn't contain the path to itself?

Reply:
Because there was no need for that.

You can derive a directory's path by tracing all the `PathSegments` in `_subdirs`.